### PR TITLE
WIP: restore order to caching git remotes

### PR DESCRIPTION
**Pull request by othiym23**
Also, update code coments to be accurate, and extirpate single-letter variable names. And introduce a params object because those argument lists were completely ridiculous.

**r:** @isaacs 
**r:** @iarna

### CODE REVIEW

File: lib/cache/add-remote-git.js [hunk 1]
```diff
@@ -26,146 +26,149 @@ var VALID_VARIABLES = [
...
+module.exports = function addRemoteGit (gitURL, silent, cb) {
+  assert(typeof gitURL === 'string', 'must have git URL')
+  assert(typeof cb === 'function', 'must have callback')
+
+  var normalized = normalizeGitUrl(gitURL)
+  var strippedURL = normalized.url
+  var treeish = normalized.branch
+  var tmpPath = strippedURL.replace(/[^a-zA-Z0-9]+/g, '-') + '-' +
+    crypto.createHash('sha1').update(strippedURL).digest('hex').slice(0, 8)
+  var checkoutPath = path.join(remotes, tmpPath)
```
Comment (iarna) [inline lib/cache/add-remote-git.js]: actual variable names!

File: lib/cache/add-remote-git.js [hunk 2]
```diff
@@ -26,146 +26,149 @@ var VALID_VARIABLES = [
...
+      ['config', '--get', 'remote.origin.url'],
+      { cwd: params.checkoutPath, env: gitEnv() },
       function (er, stdout, stderr) {
-        var stdoutTrimmed = (stdout + "\n" + stderr).trim()
-        if (er || u !== stdout.trim()) {
-          log.warn( "`git config --get remote.origin.url` returned "
-                  + "wrong result ("+u+")", stdoutTrimmed )
-          return rm(p, function (er){
+        var originURL = (stdout + '\n' + stderr).trim()
+        if (er || params.gitURL !== stdout.trim()) {
```
Comment (iarna) [inline lib/cache/add-remote-git.js]: What's going on with originURL having stderr in it, and then not using that when validating the result?

File: lib/cache/add-remote-git.js [hunk 3]
```diff
@@ -26,146 +26,149 @@ var VALID_VARIABLES = [
...
+          log.warn(
```
Comment (iarna) [inline lib/cache/add-remote-git.js]: Should the first argument be a log section eg `checkGitDir`?

File: lib/cache/add-remote-git.js [hunk 4]
```diff
@@ -173,97 +176,124 @@ function fetchRemote (p, u, co, origUrl, cb) {
...
+function resolveHead (params, cb) {
+  log.verbose('resolveHead', 'original treeish', params.treeish)
   git.whichAndExec(
-    [ "rev-list", "-n1", co ],
-    { cwd : p, env : gitEnv() },
+    ['rev-list', '-n1', params.treeish],
+    { cwd: params.checkoutPath, env: gitEnv() },
     function (er, stdout, stderr) {
-      stdout = (stdout + "\n" + stderr).trim()
+      params.exactTreeish = (stdout + '\n' + stderr).trim()
```
Comment (iarna) [inline lib/cache/add-remote-git.js]: Mutating params feels icky. And again, I'm left wondering why stderr is glommed on here?

File: lib/cache/add-remote-git.js [hunk 5]
```diff
@@ -173,97 +176,124 @@ function fetchRemote (p, u, co, origUrl, cb) {
...
-
-    mkdir(tmp, function (er) {
-      if (er) return cb(er)
-
-      git.whichAndExec(["clone", p, tmp], { cwd : p, env : gitEnv() }, clone)
-    })
-
-    function clone (er, stdout, stderr) {
-      stdout = (stdout + "\n" + stderr).trim()
+function clone (params, cb) {
```
Comment (iarna) [inline lib/cache/add-remote-git.js]: The term `clone` feels a little overloaded. `gitClone` would be clearer to me.

**Comment by iarna** [comment]
Looks great! Much easier to read. The variable name changes alone!

File: lib/cache/add-remote-git.js (hunk 4 above)
Comment (othiym23) [inline lib/cache/add-remote-git.js]: It does, but I wonder how much that is a real issue, and how much is an urge towards ideological purity, given that this particular `params` never escapes the context of this file. I have an alternative approach that I'm going to try, though, which is pulling all the non-utility parts of the caching flow back into the main function, and setting the individual variables within the scope of the closure. That's how it was originally written, and now that the control flow has been cleaned up a lot, it makes sense to bring it back to that style again.

File: lib/cache/add-remote-git.js (hunk 2 above)
Comment (othiym23) [inline lib/cache/add-remote-git.js]: That was how it was written originally, which probably means that it's a byproduct of how @isaacs originally hacked it together, and yes, I think its use is purely informational. This particular check is just verifying that the repo we've been handled is for the source we're expecting. I'll try cleaning this up and see if anything breaks.

File: lib/cache/add-remote-git.js (hunk 5 above)
Comment (othiym23) [inline lib/cache/add-remote-git.js]: I'll brainstorm this as I rewrite. More descriptive function names would probably be a good complement to the variable names.

**Comment by othiym23** [comment]
I ended up gitting some more work done on this, with more comments, gitting rid of the params object, and hopefully improving this git workflow to make it easier to git. Everyone is going to be so happy to git this change.

**Pull request merged**